### DISCUSSION TITLE
pacman `info` (Provide) `owns` & `list` executable files/commands

### DISCUSSION THREAD

**Discussion by goyalyashpal** [msys2/MSYS2-packages#3593] [Q&A] [answered]
<details>
<summary><h3>Background and Quotations</h3></summary>

there are many things:
- files (some can be executables)
- packages (collection of files)
- groups (collection of packages)

_Assumption: `-i` shows same info under "Provides" for both `-Q` and `-S`_

```
  -i, --info           view package information (-ii for backup files)
  -i, --info           view package information (-ii for extended information)
```

* a package often provides multiple executable files: 
* `-Q --list <pkg-name>` list files owned by queried package
* `-Q --owns <file>` query package that [owns] file
* `--info <package>` view package info in various fields including "Provides"

Online counterparts:
- `-Qi <pkg>`: `-Si <pkg>`
- `-Ql <pkg>`: <tt>[-Fl] \<pkg></tt> or "Files" section at `packages.msys2.org/package/<pkg>` (ref: https://github.com/msys2/MSYS2-packages/discussions/3587#discussioncomment-5112275)

</details>

[-Fl]: https://github.com/msys2/MSYS2-packages/discussions/3593#discussioncomment-5182514
[owns]: https://wiki.archlinux.org/title/Pacman#Querying_package_databases:~:text=%24%20pacman-,%2DQo,-/path/to/file_name

### Expectations

My naive understanding was that:

- `-Qi` "Provides" field ...
would show the executable files from that package 
- ... i.e. smth on likes of `-Qi <pkg> | grep -E */bin` 

<details>
<summary>Resolved</summary><br/>

- _many don't have `.exe` extension, so `-Qi <pkg> | grep --extended-regexp *.exe` won't work. `grep usr/bin` is not generic enough_
- _update: it_ is consistent _currently too. The ".exe" was significant here, which i missed._ 
- `-Ql <pkg>` and `-Qo <file>` ...
would be consistent i.e. _file_ in output from `l` when fed to `o` will result in same _pkg_

</details>

### Reality & Question

1. "Provides" field says "none" very frequently despite having the package "owning" various executable files/commands
So, what is it that "Provides" field actually lists??

2. Why do there exist 2 files for same software? i.e. one with `.exe` and other without? and they also behave different in respect to whom they are owned by (refer logs below)?
as in `more` vs `more.exe` and `cksum` vs `cksum.exe` and so on!

<details>
<summary><h3>Logs</h3><br/>Failures regarding "Provides":</summary><br/>

```console
$ # Failure: pacboy not listed
$ pacman -Qi pactoys | grep Provides
Provides        : pactoys-git  repman-git  updpkgver-git

$ pacman -Ql pactoys | grep -E */bin
pactoys /usr/bin/pacboy
   ...

$ pacman -Qo pacboy
/usr/bin/pacboy is owned by pactoys r55.8860e1f-1
```

```console
$ pacman -Qi coreutils | grep Provides
Provides        : None

$ pacman -Ql coreutils | grep -E */bin
coreutils /usr/bin/cksum.exe
coreutils /usr/bin/sha256sum.exe
   ...

$ pacman -Qo sha256sum cksum
error: No package owns /usr/bin/sha256sum
error: No package owns /usr/bin/cksum

$ # Waittt whatttt 😨 ?! these are different?
$ pacman -Qo sha256sum.exe cksum.exe
/usr/bin/sha256sum.exe is owned by coreutils 8.32-5
/usr/bin/cksum.exe is owned by coreutils 8.32-5
```

```console
$ pacboy debug -Si clang-tools-extra:c | grep Provides
Provides        : None

$ # From online counterpart to -Ql i.e. in "Files" section of ...
$ # https://packages.msys2.org/package/mingw-w64-clang-x86_64-clang-tools-extra

/clang64/bin/clang-apply-replacements.exe
/clang64/bin/clang-change-namespace.exe
/clang64/bin/clang-doc.exe
/clang64/bin/clang-include-fixer.exe
/clang64/bin/clang-move.exe
/clang64/bin/clang-pseudo.exe
/clang64/bin/clang-query.exe
/clang64/bin/clang-reorder-fields.exe
/clang64/bin/clang-tidy.exe
/clang64/bin/clangd.exe
/clang64/bin/find-all-symbols.exe
/clang64/bin/modularize.exe
/clang64/bin/pp-trace.exe
/clang64/bin/run-clang-tidy
```

```console
$ pacman -Qi inetutils | grep Provides
Provides        : None

$ pacman -Ql inetutils | grep -E */bin
inetutils /usr/bin/
inetutils /usr/bin/dnsdomainname.exe
inetutils /usr/bin/ftp.exe
inetutils /usr/bin/rcp.exe
inetutils /usr/bin/rlogin.exe
inetutils /usr/bin/rsh.exe
inetutils /usr/bin/talk.exe
inetutils /usr/bin/telnet.exe
```

</details>

**Comment by Biswa96** [comment, marked as answer]
This documentation may help https://wiki.archlinux.org/title/PKGBUILD#provides

**Reply by goyalyashpal** [reply]
thanks.

any input for the second newly added question?

**Reply by Biswa96** [reply]
> as in more vs more.exe and cksum vs cksum.exe and so on!

Where did you see more vs more.exe ?

**Reply by goyalyashpal** [reply]
in output of `pacman -Qo more.exe` and `pacman -Qo more`

```console
$ pacman -Qo more.exe
/usr/bin/more.exe is owned by util-linux 2.35.2-3

$ pacman -Qo more
error: No package owns /usr/bin/more
```

**Reply by Biswa96** [reply]
`more` does not exist. Isn't it clear from that error message?

**Reply by goyalyashpal** [reply]
no package owns is different from not exists. and `more` was expanded to `/usr/bin/more` which does give impression that it exists. Paths are not expanded for things not existing (unless its dedicated to that).

There is an issue of usage - i guess that msys2 runs commands with or without .exe BUT it actually keeps the call as it is.

Only that wouldnt be a problem - it becomes aggreviated as some msys2 commands correspond to non .exe while some do to the .exe ones.

**Reply by Biswa96** [reply]
> i guess that msys2 runs commands with or without .exe

It is not a bug, it's a feature. That feature allows to run programs without adding .exe extension. It also allows to be compatible with unix-like systems. For example, in shell script etc.

**Reply by sskras** [reply]
I'd add vice versa: 

this allows the MSYS2 tools to be run from the batch scripts.

MSYS2/Cygwin could easily strip the `.exe` from the file name and start the extensionless executable using WinAPI / `CreateProcess()`.
That would be just like a unix-like system but incompatible with the batch scripts.

@yashpalgoyal1304: if you would like to have extensionless CLI executables, I suggest you to try building @Midipix distribution:
https://github.com/lalbornoz/midipix_build#2-building-and-deployment

Only that it has no package manager at the moment. :)

**Reply by goyalyashpal** [reply]
> _"It is not a bug, it's a feature."_

i never said it was a bug, i only implied 2 things: either all of those should have .exe extension, 

or an  alternate: the pacman -Qo should work without .exe extension too

**Reply by goyalyashpal** [reply]
well never mind... i will (try to) find some regex way to supply both strings to `-Qo`

(or if u know some way to achieve that)

**Reply by sskras** [reply]
@yashpalgoyal1304, now I see where you get the extensionless path format of the command from – it's the output of `pacman -Qo`:
```
$ pacman -Ql coreutils | grep -E */bin
coreutils /usr/bin/cksum.exe
coreutils /usr/bin/sha256sum.exe
   ...

$ pacman -Qo sha256sum cksum
error: No package owns /usr/bin/sha256sum
error: No package owns /usr/bin/cksum
```

To developers: perhaps `pacman -Qo ...` should add `.exe` automatically when the arguments are provided without a path (and when the found path is added to the file automatically), umm?

**Reply by goyalyashpal** [reply]
> perhaps pacman -Qo ... should add .exe automatically when the arguments are provided without a path

yes, that would be awesome, i guess 😅 . that is, if its not considered bloat 😆 by others

**Reply by Biswa96** [reply]
I think the behavior is expected. /usr/bin/sha256sum really does not exist. So, pacman is right that `/usr/bin/sha256sum` is not owned by any package. If you want to search file use `pacman -F` command.

> should add .exe automatically when the arguments are provided without a path

Adding .exe automatically would create issues for the packages which contains both foo and foo.exe files. For example, mingw-w64-mpv

```
$ pacman -Qo /ucrt64/bin/mpv
/ucrt64/bin/mpv is owned by mingw-w64-ucrt-x86_64-mpv 0.35.1-3
$ pacman -Qo /ucrt64/bin/mpv.exe
/ucrt64/bin/mpv.exe is owned by mingw-w64-ucrt-x86_64-mpv 0.35.1-3
```

**Reply by goyalyashpal** [reply]
> Adding .exe automatically would create issues for the packages which contains both foo and foo.exe files. For example, mingw-w64-mpv

thats why i was thinking of some way to supply both names at once via some external means (possibly using regex)

**Reply by sskras** [reply]
@Biswa96 wrote:

> packages which contains both foo and foo.exe files. For example, mingw-w64-mpv
> 
> ```
> $ pacman -Qo /ucrt64/bin/mpv
> /ucrt64/bin/mpv is owned by mingw-w64-ucrt-x86_64-mpv 0.35.1-3
> $ pacman -Qo /ucrt64/bin/mpv.exe
> /ucrt64/bin/mpv.exe is owned by mingw-w64-ucrt-x86_64-mpv 0.35.1-3
> ```

What is the reason to add two executables along each other?

Heck, this is an ugly hack!
https://github.com/msys2/MINGW-packages/commit/6f2504942c91b1d15878fa1de6e3e057a93d4d75#diff-c433d4797ff26bbf54e02123bb2b5e1df3e652bdcb811865bbf4e9188a0049ddR92

And it's not going away anytime soon, I guess:
```
package() {
  ...
  # mpv needs winpty for key bindings to work on the terminal
  mv "${pkgdir}${MINGW_PREFIX}/bin/mpv.exe" "${pkgdir}${MINGW_PREFIX}/bin/mpv_exe"
  _exename=mpv
  ...
  echo '/usr/bin/winpty ... >> "${pkgdir}${MINGW_PREFIX}/bin/${_exename}"
  mv "${pkgdir}${MINGW_PREFIX}/bin/mpv_exe" "${pkgdir}${MINGW_PREFIX}/bin/mpv.exe"
}
```

That's not a clean (and even ugly) way to do the integration.

---

Another good point to consider while developing the @midipix (which doesn't use `.exe` extensions).

**Comment by sskras** [comment]
@yashpalgoyal1304 wrote [2 hours ago](https://github.com/msys2/MSYS2-packages/discussions/3593#discussioncomment-5177208):

> thats why i was thinking of some way to supply both names at once via some external means (possibly using regex)

I still don't understand your primary aim. Do you wan to extract all the executable names (and skip other thing like DLLs)?

In the OP you use this following syntax:
```
... | grep -E */bin
```

The leading `*` is useless. It's the operator meaning "the preceding symbol occurs zero or more times". And there you give no preceding symbols here. So it becomes just `grep /bin`:
```
$ pacman -Ql mingw-w64-x86_64-mpv | grep /bin
mingw-w64-x86_64-mpv /mingw64/bin/
mingw-w64-x86_64-mpv /mingw64/bin/libmpv-2.dll
mingw-w64-x86_64-mpv /mingw64/bin/mpv
mingw-w64-x86_64-mpv /mingw64/bin/mpv.com
mingw-w64-x86_64-mpv /mingw64/bin/mpv.exe
```

Now alongside `bin` there might be binaries in `sbin` subdir. Name of both subdirs starts differently but ends the same.
So it needs the slash moved to the end of `...bin` instead, hence `grep bin/`, and in our case nothing changes:
```
$ pacman -Ql mingw-w64-x86_64-mpv | grep bin/
mingw-w64-x86_64-mpv /mingw64/bin/
mingw-w64-x86_64-mpv /mingw64/bin/libmpv-2.dll
mingw-w64-x86_64-mpv /mingw64/bin/mpv
mingw-w64-x86_64-mpv /mingw64/bin/mpv.com
mingw-w64-x86_64-mpv /mingw64/bin/mpv.exe
```

Now name of the extensionless binary can start by any char except for the `.` symbol, right?, hence `[^.]`:
```
$ pacman -Ql mingw-w64-x86_64-mpv | grep bin/[^.]
mingw-w64-x86_64-mpv /mingw64/bin/libmpv-2.dll
mingw-w64-x86_64-mpv /mingw64/bin/mpv
mingw-w64-x86_64-mpv /mingw64/bin/mpv.com
mingw-w64-x86_64-mpv /mingw64/bin/mpv.exe
```

And there should be at least one such char before the end of the string (the extensionless name), which in RegExps is denoted by the `$`.
This implies the operator `+` in the extended RE syntax, so we return `-E` switch back:
```
$ pacman -Ql mingw-w64-x86_64-mpv | grep -E bin/[^.]+$
mingw-w64-x86_64-mpv /mingw64/bin/mpv
```

And finally we filter out only the wanted extension that we count as executable, I suppose the that's `.exe`.
For that we should escape the `.` symbol to become `\.` (otherwise it means any possible symbol in the string).
So the part or RexExp to match this extension becomes `\.exe`, which we expect just before the end of the line, the `$`:
```
$ pacman -Ql mingw-w64-x86_64-mpv | grep -E 'bin/[^.]+\.exe$'
mingw-w64-x86_64-mpv /mingw64/bin/mpv.exe
```

Now we might make matching the file extension optional by embracing it using braces, `(\.exe)`, and putting the `?` operator after it which means "occurs zero or one time":
```
$ pacman -Ql mingw-w64-x86_64-mpv | grep -E 'bin/[^.]+(\.exe)?$'
mingw-w64-x86_64-mpv /mingw64/bin/mpv
mingw-w64-x86_64-mpv /mingw64/bin/mpv.exe
```

And if you would like to include some other file extension, eg. `.com`, you can use the RegExp OR-operator denoted by `|` symbol so that part becomes `(\.com|\.exe)`:
```
$ pacman -Ql mingw-w64-x86_64-mpv | grep -E 'bin/[^.]+(\.exe|\.com)?$'
mingw-w64-x86_64-mpv /mingw64/bin/mpv
mingw-w64-x86_64-mpv /mingw64/bin/mpv.com
mingw-w64-x86_64-mpv /mingw64/bin/mpv.exe
```

Is that something like what you wanted?

**Reply by goyalyashpal** [reply]
thanks a lot for the thorough walkthrough about usage of regex. i had tried learning about regex 3 times in my life by watching videos or such, never got it - as those videos focused on teaching it in imaginary situations.

ur this hands on walkthrough will help me alot.

as for what i wanted, i will try to experiment and come back later with my input.

**Reply by goyalyashpal** [reply]
yeah so, the summary: 
1. feed the strings into `pacman -Qo` anyhow
2. generate two different strings programmatically anyhow (i.e. regex not compulsory). 
say "cksum" and "cksum.exe" using "cksum" and "exe"
3. extension of or other way to put 2. can be that generate a _[cartesian product]_ from 2 sets of strings so, _mathematically_ writing:
`{ "cksum", "sha256sum" } ⨯ { "", "exe" }` to get `{ "cksum", "cksum.exe", "sha256sum", "sha256sum.exe" }`


---
### Attempts:
- 1\. can be achieved fairly easily: using `-` as argument and `<<<` for reading standard input. so: 

```console
$ # Enclosing in quotes treats as single string
$ pacman -Qo - <<< cksum sha256sum
error: No package owns /usr/bin/sha256sum
error: No package owns /usr/bin/cksum
```

- 2\. can be achieved using cartesian product via [bash brace expansion]

```console
$ # The e is any dummy character to trigger the expansion, couldn't figure out how otherwise
$ pacman -Qo - <<< e {cksum,sha256sum}{.exe,}
/usr/bin/cksum.exe is owned by coreutils 8.32-5
error: No package owns /usr/bin/cksum
/usr/bin/sha256sum.exe is owned by coreutils 8.32-5
error: No package owns /usr/bin/sha256sum
error: No package owns e
```

- possible alternate solutions for above using: [perl5] (pacman-contrib dependency) | [perl6] | [zsh solution][zsh-sol]
[(inverse) regex data generator][text-from-regex] are more suitable for random data generation, not this.

[cartesian product]: https://codegolf.stackexchange.com/questions/195964/#comment466287_195964 "n-dimensional-cartesian-product"
[perl5]: https://codegolf.stackexchange.com/a/165678/112922 "cartesian-product-of-a-list-with-itself-n-times"
[perl6]: https://codegolf.stackexchange.com/a/195999/112922 "n-dimensional-cartesian-product"
[zsh-sol]: https://codegolf.stackexchange.com/a/195984/112922 "n-dimensional-cartesian-product"
[bash brace expansion]: https://codegolf.stackexchange.com/a/143546/112922 "bash brace expansions"
[text-from-regex]: https://www.browserling.com/tools/text-from-regex "text-from-regex"








[using regex to generate strings]: https://duckduckgo.com/?q=using+regex+to+generate+strings&t=vivaldi&ia=web
[strings-from-regex]: https://onlinestringtools.com/generate-string-from-regex "Based on browserling's tool"
[generate-rather-than-match]: https://stackoverflow.com/questions/22115/#22133 "using-regex-to-generate-strings-rather-than-match-them"

[pairwise generation]: https://stackoverflow.com/a/1015626/8395964 "case generator from combinations of sets"
[iterator.product]: https://codegolf.stackexchange.com/questions/195964/#comment466268_195964 "n-dimensional-cartesian-product"

**Reply by goyalyashpal** [reply]
### Further experiments:

```console
$ # Aim: filter lines for "owned"
$ pacman -Qo - <<< e {cksum,sha256sum}{.exe,} | grep owned
error: No package owns /usr/bin/cksum
error: No package owns /usr/bin/sha256sum
error: No package owns e
/usr/bin/cksum.exe is owned by coreutils 8.32-5
/usr/bin/sha256sum.exe is owned by coreutils 8.32-5

$ # FAIL as well as PASS
$ # Aim failed BUT "error:" and successful results are sorted, i don't know how 😅
$ # wowwww, i wanted to do this too, it's looking beautiful. 
$ # It's more like a "syntax error" created a feature lol
```

```console
$ # Aim: Don't mix input and output on same side
$ {cksum,sha256sum}{.exe,} | pacman -Qo -
$ # FAIL: First argument is not just a string anymore, acts as command

$ # Aim: Make it as command on left side
$ echo {cksum,sha256sum}{.exe,} | pacman -Qo -
error: No package owns cksum.exe cksum sha256sum.exe sha256sum
$ # FAIL: Output of echo becomes one string, not multiple
```

**Reply by goyalyashpal** [reply]
### Further experiments with 1st experiment above:

```console
$  pacman -Qo - <<< e {cksum,sha256sum}{.exe,} | grep owned
error: No package owns /usr/bin/cksum
error: No package owns /usr/bin/sha256sum
error: No package owns e
/usr/bin/cksum.exe is owned by coreutils 8.32-5
/usr/bin/sha256sum.exe is owned by coreutils 8.32-5

$  pacman -Qo - <<< e {cksum,sha256sum}{.exe,} | grep by
error: No package owns /usr/bin/cksum
error: No package owns /usr/bin/sha256sum
error: No package owns e
/usr/bin/cksum.exe is owned by coreutils 8.32-5
/usr/bin/sha256sum.exe is owned by coreutils 8.32-5

$  pacman -Qo - <<< e {cksum,sha256sum}{.exe,} | grep package
error: No package owns /usr/bin/cksum
error: No package owns /usr/bin/sha256sum
error: No package owns e
```

**Reply by sskras** [reply]
What are you trying to achieve? What's the goal?

**Reply by sskras** [reply]
If I would need to get the executable names known to pacman, I would do this:
```
$ which {cksum,sha256sum}{.exe,} | pacman -Qo - |& grep -v ^error
/usr/bin/cksum.exe is owned by coreutils 8.32-4
/usr/bin/sha256sum.exe is owned by coreutils 8.32-4
```

... or maybe `which` should be replaced with `command -v` for the performance reasons (to avoid `fork()` syscall which is slow on Windows).

**Reply by sskras** [reply]
PS. Should I need only the names that exists on disk, this would do:
```
$ command -v {cksum,sha256sum}{.exe,} | pacman -Qo - |& awk '/^\// {print $1}'
/usr/bin/cksum.exe
/usr/bin/sha256sum.exe
```

**Reply by goyalyashpal** [reply]
> What are you trying to achieve? What's the goal?

yes, this one:

> get the executable names known to pacman

```console
$ command -v {cksum,sha256sum}{.exe,} | pacman -Qo - |& grep -v ^error
/usr/bin/cksum.exe is owned by coreutils 8.32-5
/usr/bin/sha256sum.exe is owned by coreutils 8.32-5

$ command -v more{.exe,} | pacman -Qo - |& grep -v ^error
/usr/bin/more.exe is owned by util-linux 2.35.2-3
```



The suggested answer worked super well. thanks for it and the addtional tip below too :)

**Reply by Biswa96** [reply]
Does the following command look more complex than the above?

```
$ pacman -F more.exe
msys/util-linux 2.35.2-3 [installed]
    usr/bin/more.exe
```

**Reply by sskras** [reply]
I still don't like that you need to concatenate input commands in this specific format. This looks cleaner to me:
```
$ CMDS="cksum sha256sum"
$ CMD_ALIASES=`echo $CMDS | xargs -n1 | while read; do echo $REPLY{.exe,}; done | xargs`
$ command -v $CMD_ALIASES | pacman -Qo - |& awk '/^\// {print $1}'
/usr/bin/cksum.exe
/usr/bin/sha256sum.exe
```

I broke the whole long command line down into several on purpose / for easier reading.

**Reply by goyalyashpal** [reply]
> Does the following command look more complex than the above?
> `pacman -F more.exe`

yes, if i have to try multiple combinations of whether the .exe is present or not.... and then i often forget to cover all combination cases when they have to be split across different commands 

one command; be it longer; if it covers more cases (i.e. hence is more fool proof for me), works much better for me.

**Reply by sskras** [reply]
@Biswa96:

> Does the following command look more complex than the above?
> 
> ```
> $ pacman -F more.exe
> msys/util-linux 2.35.2-3 [installed]
>     usr/bin/more.exe
> ```

Yes, it needs additional work:
```
$ pacman -F more.exe
warning: database file for 'mingw32' does not exist (use '-Fy' to download)
warning: database file for 'mingw64' does not exist (use '-Fy' to download)
warning: database file for 'ucrt64' does not exist (use '-Fy' to download)
warning: database file for 'clang32' does not exist (use '-Fy' to download)
warning: database file for 'clang64' does not exist (use '-Fy' to download)
warning: database file for 'msys' does not exist (use '-Fy' to download)
```

**Reply by sskras** [reply]
@goyalyashpal wrote:

> one command; be it longer; if it covers more cases (i.e. hence is more fool proof for me), works much better for me.

Thanks to you I now have additional function in my Bash environment:
```
$ type realbins
realbins is a function
realbins ()
{
    CMD_ALIASES=`echo $* | xargs -n1 | while read; do echo $REPLY{.exe,}; done | xargs`;
    command -v $CMD_ALIASES | pacman -Qo - 2>&1 | awk '/^\// {print $1}'
}

$ realbins mpv
/mingw64/bin/mpv.exe
/mingw64/bin/mpv

$ realbins cksum sha256sum
/usr/bin/cksum.exe
/usr/bin/sha256sum.exe
```

**Reply by goyalyashpal** [reply]
_THIS COMMENT IS FOR MY OWN_
_FEEL FREE TO IGNORE :)_

----
> `... |& grep ...`
> \- Abridged from @ sskras at https://github.com/msys2/MSYS2-packages/discussions/3593#discussioncomment-5182365



The <tt>[|&]</tt>[^unix.SE] redirection operator sends both stdout & stderr

[^unix.SE]: Article linked at [what-is-symbol-and-in-unix-linux] <-- ddg :mag:: [|& in bash]

Ahw:
- `-Qo` gives errors to stderr (i.e. already sorted, nice)
- plain pipe `|` redirects just stdout

* This explains earlier experiment resuls, 
why `... | grep ...` _"sorted"_ results: errors outputted, stdout redirected

So:
- `pacman -Qo - <<< e {cksum,sha256sum}{.exe,} |& grep by` works too, although is hakish
- some way to "discard" or [suppress-stderr][how-to-suppress-stdout-and-stderr-in-bash] stream will be suffice too (no need to `grep`)

<details>
<summary><tt>More experimentations</tt></summary>

```console
$ # Old Proper way
$ command -v {cksum,sha256sum}{.exe,} | pacman -Qo - |& grep -v ^error

$ # Old hack, modified
$ pacman -Qo - <<< e {cksum,sha256sum}{.exe,} |& grep by
```


```console
$ # New proper way: suppress stderr of -Qo
$ pacman -Qo - <<< $(command -v {cksum,sha256sum}{.exe,}) 2> /dev/null
$ # Yes, worked! This even maintains colors lol :)

$ # Using part of old hack to shorten a bit
$ pacman -Qo - <<< e {cksum,sha256sum}{.exe,} 2> /dev/null
```

</details>

```console
$ # Let's try removing input redirection 😋 
$ pacman -Qo {cksum,sha256sum}{.exe,} 2> /dev/null
/usr/bin/cksum.exe is owned by coreutils 8.32-5
/usr/bin/sha256sum.exe is owned by coreutils 8.32-5

$ # Woww! Short, Colourful, Nice :)
```


[|&]: https://catonmat.net/bash-one-liners-explained-part-three "CatOnMat article"
[|&]: https://web.archive.org/web/20221224120924/https://catonmat.net/bash-one-liners-explained-part-three
[what-is-symbol-and-in-unix-linux]: https://unix.stackexchange.com/a/89388/448123 "Link in Unix.SE answer"
[|& in bash]: https://duckduckgo.com/?q=%7C%26+in+bash&t=vivaldi&ia=web "ddg search"
[how-to-suppress-stdout-and-stderr-in-bash]: https://stackoverflow.com/q/46009283/8395964 "SO question"

---
_**Update:** Following not needed anymore_
Other is that, rearranging the input strings part towards right will help in reusing it for different input with up arrows.

`<filtering stderr> <<< pacman -Qo - <<< e {cksum,sha256sum}{.exe,}`

**Reply by sskras** [reply]
@goyalyashpal: 

With all due respect, at least half of your text is quite unrelated to MSYS2.
Maybe you should start your blog about learning basics of *nix CLI?

You could even host it here, on GitHub – no need to buy a domain name: https://pages.github.com/

**Reply by goyalyashpal** [reply]
> Maybe you should start your blog about learning basics of *nix CLI?

u r exactly on point. i was thinking about that from long time, but couldn't get time.

**Reply by goyalyashpal** [reply]
> `$ pacman -F more.exe`
\- @ Biswa96 at https://github.com/msys2/MSYS2-packages/discussions/3593#discussioncomment-5182514

```console
$  pacman -F more{,.exe}
msys/util-linux 2.35.2-3 [installed]
    usr/share/bash-completion/completions/more
msys/util-linux 2.35.2-3 [installed]
    usr/bin/more.exe
```

Thanks for sharing that. 
- it ain't better for querying _local_ packages ('ts slow, and overkill)
- But it's a must-have companion to `-Ss` for finding remote package name 
when the file name is not same as package name.
- I was always searching for smth like this (e.g. when i had to find [pacboy], & [clangd])

> To query the database to know which remote package a file belongs to:
>\- https://wiki.archlinux.org/title/Pacman

[pacboy]: https://github.com/msys2/MSYS2-packages/issues/3457#issuecomment-1368204250
[clangd]: https://github.com/clangd/clangd/discussions/1170#discussioncomment-5158727

So:

```console
$ pacman -Ss pacboy

$ pacman -F pacboy{,.exe}
msys/pactoys r55.8860e1f-1 [installed]
    usr/bin/pacboy
    usr/share/bash-completion/completions/pacboy

$ pacboy -Ss clangd

$ pacman -F clangd{,.exe}
clangarm64/mingw-w64-clang-aarch64-clang-tools-extra 15.0.7-3 (mingw-w64-clang-aarch64-toolchain)
    clangarm64/bin/clangd.exe
   ...
```

----

[remove package database]: https://www.reddit.com/r/archlinux/comments/hv3njn/comment/fyqzl7j/ "Remove package database from `-Fy`"
[ddg how to delete pacman database file]: https://duckduckgo.com/?q=how+to+delete+pacman+database+file&t=vivaldi&ia=web

Here's how to [remove package database]. I verified it with doublecmd open at that path while doing `-Fy`.

<details>
<summary><tt>-Fy logs</tt></summary>

```console
$ pacman -F clangd
warning: database file for 'clangarm64' does not exist (use '-Fy' to download)
   ...

$ realpath $(pacman-conf dbpath)/sync/*.files
/var/lib/pacman/sync/*.files

$ realpath --help
Usage: realpath [OPTION]... FILE...
Print the resolved absolute file name;
all but the last component must exist
   ...

$ pacman -Fy
:: Synchronizing package databases...
   ...

$ realpath $(pacman-conf dbpath)/sync/*.files
/var/lib/pacman/sync/clang32.files
   ...
```

</details>
